Fix #331 "We only get `Const::Num`, never `Const::Int`"

HalidOdat opened the pull request:
This will close #331 and unblock #328

jasonwilliams commented:
Thanks, will try and get round to looking at this soon

Razican commented on boa/src/syntax/lexer/mod.rs:
```diff
@@ -399,16 +399,18 @@ impl<'a> Lexer<'a> {
                                     }
                                 }
                             }
+                            dbg!(&gone_decimal);
```
This should be removed before merging.

Razican commented on boa/src/syntax/lexer/mod.rs:
```diff
@@ -399,16 +399,18 @@ impl<'a> Lexer<'a> {
                             } else {
-                                (u64::from_str_radix(&buf, 8).map_err(|_e| LexerError::new("Could not convert value to u64"))?) as f64
+                                NumericLiteral::Integer(
+                                u64::from_str_radix(&buf, 8).map_err(|_e| {dbg!(&_e); LexerError::new("Could not convert value to u64")})?)
```
The `dbg!()` call should be removed before merging.
```suggestion
                                u64::from_str_radix(&buf, 8).map_err(|_e| LexerError::new("Could not convert value to u64"))?)
```

Razican commented on boa/src/syntax/lexer/mod.rs:
```diff
@@ -469,9 +474,19 @@ impl<'a> Lexer<'a> {
+                            )
+                        } else {
+                            NumericLiteral::Integer(
+                                u64::from_str(&buf).map_err(|_e| {dbg!(&_e); dbg!(&is_floating_point); LexerError::new("Could not convert value to u64")})?
```
The `dbg!()` macro should be removed before merging.
```suggestion
                                u64::from_str(&buf).map_err(|_e| LexerError::new("Could not convert value to u64"))?
```

Razican commented on boa/src/syntax/ast/token.rs:
```diff
@@ -42,6 +42,17 @@ impl Debug for VecToken {
+#[derive(Clone, Copy, PartialEq, Debug)]
+pub enum NumericLiteral {
+    /// A floating point number
+    Number(f64),
```
I would probably change this `Number` for `Real` or something like that, to differentiate from an `Integer`.

HalidOdat replied:
In V8 they have `Number` and `Smi`, so should I do it like V8 or `Real` and `Integer`?

Razican replied:
> In V8 they have `Number` and `Smi`, so should I do it like V8 or `Real` and `Integer`?

I think it's more accurate if we call `Real` or even `Rational` to non-integers.

Razican commented on boa/src/syntax/ast/token.rs:
```diff
@@ -42,6 +42,17 @@ impl Debug for VecToken {
+    Number(f64),
+    /// A integer
+    Integer(u64),
+    // TODO: Add BigInt
```
How big is a `BigInt`?

HalidOdat replied:
BigInt can hold any arbitrary large integer. All the numerical are implemented in the software level.

[num-bigint](https://crates.io/crates/num-bigint) crate

HalidOdat commented:
I reworked the number lexing and fixed some things:
 - Added some documentation to `lexer` and `token`
 - Fixed a bug were it did not accept `09.5e2`.
 - Made it easy to add `strict mode` for number literal (in strict mode `034` and `098` is an error).
 - Made it easy to add `BigInt` support (including other base BigInts).
 - Added `Value::is_integer` to check if a number is an integer.
 - Renamed `ValueData::Number(f64)` to `ValueData::Racional(f64)`
 - Fixed position bug related to numeric tokens.
 - Added some tests for edge case with implicit octal and hexadecimals followed by a dot
 - Split numerical lexing to a different function.
 - Removed `read_integer_in_base`.
 
This **unblocks** #328 

I think now it's ready to review. :)

Razican commented on boa/src/builtins/value/mod.rs:
```diff
@@ -43,7 +43,7 @@ pub enum ValueData {
     String(String),
     /// `Number` - A 64-bit floating point number, such as `3.1415`
-    Number(f64),
+    Racional(f64),
```
Why is this `Racional` instead of `Rational`?

HalidOdat replied:
> Why is this `Racional` instead of `Rational`?

Good catch! I wouldn't have noticed if I wasn't concentrated on it. It's because in my native language that's
how you spell `Rational` in my mind it didn't seem wrong at the time.

You're right it should be `Rational`. :)

Razican replied:
It happens to me too. In Spanish it's spelled "racional" too haha

Razican commented on boa/src/syntax/lexer/tests.rs:
```diff
@@ -336,7 +336,6 @@ fn test_two_divisions_in_expression() {
     let s = "    return a !== 0 || 1 / a === 1 / b;";
     let mut lexer = Lexer::new(s);
     lexer.lex().expect("failed to lex");
-    dbg!(&lexer.tokens);
```
This `dbg!` here was to be able to debug the token generation from the lexer when we activate this test (when we solve the goal-oriented lexer). We can remove it, but not sure if there is a need.

Razican commented on boa/src/syntax/ast/token.rs:
```diff
@@ -42,28 +46,53 @@ impl Debug for VecToken {
+#[derive(Clone, Copy, PartialEq, Debug)]
+pub enum NumericLiteral {
+    /// A floating point number
+    Racional(f64),
```
Again, isn't the correct word `Rational`?

Razican commented on boa/src/syntax/lexer/tests.rs:
```diff
@@ -443,55 +541,88 @@ fn test_regex_literal_flags() {
     assert_eq!(
         lexer.tokens[2].kind,
-        TokenKind::NumericLiteral(100_000_000_000.0)
+        TokenKind::NumericLiteral(NumericLiteral::Racional(100_000_000_000.0))
```
If we start seeing this a bit too wordy, we could implement `From<f64>` and `From<i32>` for `NumericLiteral`, and then add a `fn numeric_literal<L>(lit: L) -> Self where L: Into<NumericLiteral> {}` to `TokenKind`, which would directly allow us to pass a `f64` or a `i32`. I think I already did something similar in the modular parser PR.

HalidOdat commented on boa/src/syntax/lexer/mod.rs:
```diff
@@ -229,6 +223,225 @@ impl<'a> Lexer<'a> {
+            }
+        }
+
+        // TODO: Setup strict mode.
+        let strict_mode = false;
```
Currently I have `strict_mode` set to always be `false`. In strict mode decimals with leading zeros and implicit octal are not allowed.

I seems like the parser should take in a lexer, because `'use strict';` it only applies to function or program scope and the lexer does not know what a scope is.

Razican replied:
Yep, this is going to be interesting to see once we have the lexer as an iterator tied to the cursor.

Razican commented:
I would wait for #304 to land before this.

HalidOdat commented:
I will rebase today!
